### ISSUE TITLE
ROCm consumer GPU (gfx1100/RDNA3): fused-MoE blocked by upstream triton AMD backend bug — request configs, allowlist fix, and mitigation

### ISSUE THREAD

**Issue by AIwork4me** [sgl-project/sglang#30245]
### Summary

On AMD consumer RDNA3 GPUs (gfx1100: RX 7900 XT/XTX, Radeon PRO W7900), SGLang cannot serve any MoE model through the `fused_moe_triton` path. The server boots, loads weights, allocates KV cache, and reaches HTTP ready — but crashes on the first MoE forward with a GPU memory access fault caused by an **upstream triton AMD backend codegen bug** (reported: triton-lang/triton#10808).

This issue requests: (a) AMD benchmark configs for common MoE shapes, (b) gfx1100 allowlist fix in `sgl-kernel/setup_rocm.py`, (c) documentation of the known limitation, and (d) exploration of a possible PyTorch-native fallback for the MoE runner.

### Environment

| Item | Detail |
|------|--------|
| GPU | AMD Radeon PRO W7900 × 4 (gfx1100, RDNA3), 48 GB each |
| ROCm | Driver 7.2.1, userland torch 2.5.1+rocm6.2 (backward-compat mode) |
| Triton | 3.1.0 (bundled with torch 2.5.1+rocm6.2) |
| SGLang | 0.0.0.dev11416+g92e8bb79e |
| Model | `baidu/Unlimited-OCR` (DeepseekV2-style MoE, E=64 experts, N=896 hidden, BF16, `UnlimitedOCRForCausalLM`) |
| Python | 3.12.3 |

### The Crash

Server boot sequence:
1. Config + tokenizer + remote-code load → **OK**
2. Weight load → **OK** (`UnlimitedOCRForCausalLM`, 6.30 GB)
3. KV cache alloc → **OK** (31.89 GB)
4. HTTP listener → **OK** ("fired up and ready to roll", `/health` → 200)
5. **First MoE forward → FAULT**

```
Memory access fault by GPU node-2 (Agent handle: 0x4a269150) on address 
0x7efe3fa50000. Reason: Page not present or supervisor privilege.
Fatal Python error: Aborted
```

Call stack:
```
triton/backends/amd/compiler.py:261 in hash          ← AMD backend JIT
triton/compiler/compiler.py:240 in compile
...fused_moe_triton/fused_moe_triton_kernels.py:1005 in act_and_mul_triton
...fused_moe_triton/fused_moe.py:527 in fused_experts_impl
```

The crash is in triton's AMD backend JIT compiler — the generated GPU code object contains an invalid memory reference. **This is an upstream triton-lang/triton bug, not an SGLang code defect.** SGLang is the trigger, not the root cause. Upstream: https://github.com/triton-lang/triton/issues/10808

### Root Cause Analysis

The crash involves two independent issues:

#### Issue A (Performance): No AMD fused-MoE benchmark config for E=64,N=896

The `fused_moe_triton/configs/` directory has pre-tuned configs for this MoE shape for many NVIDIA GPUs (H100, B200, H200, A100, A800, RTX 4090) but **zero for any AMD GPU** — not even datacenter MI300X/MI325X.

AMD configs exist only for triton 3.1.0 and only for these shapes:
- E=8, N=1792/3584/4096/7168/8192/14336 (DeepSeek shared-expert style)
- E=256, N=256, fp8 only

The missing config file would be: `E=64,N=896,device_name=AMD_Radeon_Graphics.json`

**This alone does NOT cause the crash** — the default config fallback produces correct (though suboptimal) kernels. The crash is a separate codegen bug.

#### Issue B (Correctness): triton AMD backend JIT produces invalid GPU code on gfx1100

The `fused_moe_kernel` (`@triton.jit` decorated) is JIT-compiled by triton's AMD backend. On gfx1100, the compiled code object page-faults at first execution. The crash location in `compiler.py:261 hash` (code object caching/loading) suggests the generated binary references unmapped GPU virtual addresses — likely a register allocation or address computation bug in the AMD backend codegen, triggered by gfx1100's RDNA3 ISA differences from gfx942 (CDNA3).

Upstream triton issue: https://github.com/triton-lang/triton/issues/10808

#### Why no alternative MoE backend works on gfx1100

`MoeRunnerBackend` options and status on this ROCm consumer GPU:

| Backend | Status |
|---------|--------|
| TRITON (default) | **Crash** — upstream triton AMD backend bug |
| TRITON_KERNELS | Requires external `triton_kernels` package (not installed in ROCm) |
| FLASHINFER_TRTLLM | NVIDIA-only (needs libcudart) |
| FLASHINFER_CUTLASS | NVIDIA-only |
| FLASHINFER_CUTEDSL | NVIDIA-only |
| FLASHINFER_MXFP4 | NVIDIA-only (SM100) |
| DEEP_GEMM | NVIDIA-only (CUTLASS-based, SM90+) |
| CUTLASS | NVIDIA-only |
| MARLIN | Not installed / NVIDIA-only |
| AITER | ROCm library, but no pre-built gfx1100 wheel for torch 2.5.1; SGLang import path requires ~40 stub import sites |

### Additional: sgl-kernel gfx1100 allowlist gap

`sgl-kernel/setup_rocm.py` line 75 only permits `gfx942` and `gfx950` for HIP compilation. gfx1100 is excluded, so even once triton's bug is fixed, `sgl-kernel` native fused ops (`silu_and_mul`, `gelu_and_mul`, `moe_align_block_size`, etc.) cannot be built for consumer RDNA3 GPUs.

Suggested patch:
```python
# Line 75: add gfx1100
if amdgpu_target not in ["gfx942", "gfx950", "gfx1100"]:

# Line 89: gfx1100 uses same 48KB smem budget as gfx942
topk_dynamic_smem_bytes = 48 * 1024 if amdgpu_target in ("gfx942", "gfx1100") else 32 * 1024 * 4
```

### Reproduction

```bash
# 1. Create venv (exact recipe: scripts/install_sglang_serve_venv.sh)
#    Key: torch 2.5.1+rocm6.2, SGLang --no-deps, skip torchao/[all_hip],
#    stub aiter for import resolution, unpack prebuilt sgl-kernel .so

# 2. Launch server
python -m sglang.launch_server \
  --host 127.0.0.1 --port 30000 \
  --model baidu/Unlimited-OCR --revision 84757cb0 --trust-remote-code \
  --dtype bfloat16 --context-length 32768 \
  --attention-backend triton --page-size 1 --mem-fraction-static 0.8 \
  --enable-custom-logit-processor --disable-overlap-schedule \
  --disable-cuda-graph --skip-server-warmup

# 3. Server reaches "ready" (/health → 200, /model_info → 200)
# 4. Send any /v1/chat/completions → scheduler hangs → crash
```

Flag deviations from NVIDIA recipe (documented for AMD):
- `--attention-backend triton` (not `fa3`: asserts NVIDIA SM 80/90)
- `--disable-cuda-graph` (hangs on capture on gfx1100)
- `--skip-server-warmup` (SIGABRT on warmup forward)

Full crash analysis + reproduction: https://github.com/AIwork4me/Unlimited-OCR-ROCm/blob/main/docs/upstream/sglang-rocm-enablement.md

### Requests

1. **Add AMD fused-MoE benchmark configs** for E=64,N=896 and other common DeepseekV2-style shapes, at least for triton 3.1.0 (`device_name=AMD_Radeon_Graphics` and `AMD_Instinct_MI300X`)

2. **Add gfx1100 to `sgl-kernel/setup_rocm.py` allowlist** — 3-line patch with same 48KB smem budget as gfx942 (see patch above)

3. **Document the known limitation** in `docs/platforms/amd_gpu.md`:
   - Consumer RDNA3 (gfx1100) cannot serve MoE models through `fused_moe_triton` due to upstream triton AMD backend bug (triton-lang/triton#10808)
   - Link to upstream triton issue for tracking
   - Suggest workaround if available

4. **Explore mitigation**: is there a way to force the MoE runner to use a PyTorch-native GEMM path (`F.linear` + `F.silu`) instead of the fused triton kernel? This would be slow but correct for consumer GPU users until triton is fixed.

**Comment by WhatGhost** [comment]
@AIwork4me  Hi — I ran some repros on gfx1100 (RDNA3) with the same setup (E=64, N=896, BF16). The MoE triton kernels themselves compile and run fine for me on both triton 3.1.0 and 3.5.1 — no fault, correct results.

But I was able to reproduce the exact same crash by feeding out-of-range expert_ids into fused_moe_kernel: it does an unmasked b = tl.load(b_ptr + off_experts * stride_be + ...), so a garbage/OOB off_experts reads unmapped VRAM and gives exactly your:
```Memory access fault ... Reason: Page not present or supervisor privilege```

Those expert_ids come from moe_align_block_size (allocated with torch.empty and filled by the sgl_kernel op, which isn't built for gfx1100 by default). Could you check whether the expert_ids produced by moe_align_block_size are as expected (all within [0, num_experts))?

**Comment by AIwork4me** [comment]
@WhatGhost — confirmed on all three points, verified empirically on our gfx1100 box. **The triton fused-MoE kernel is correct on gfx1100; the crash is exactly the OOB-`expert_ids` → unmasked `tl.load` you described; and `moe_align_block_size` does emit garbage `expert_ids` here.** The root cause sits one layer below `moe_align`: the only `sgl_kernel` build available for gfx1100 is a **gfx942 binary**, so the whole MoE-gating/activation path is non-functional. Our original "triton AMD backend codegen bug" attribution in the issue body was a misdiagnosis — `triton-lang/triton#10808` is not implicated.

## Setup & what we did

Host: **gfx1100 (RDNA3)**, ROCm 6.2 / torch 2.5.1+rocm6.2, triton 3.1.0, sglang `dev11416`. We drove the **real** `fused_experts_impl` triton path end-to-end and controlled exactly one variable — the `expert_ids` tensor — by monkeypatching `fused_moe.moe_align_block_size`. We also called `sgl_kernel.moe_align_block_size` directly to inspect its raw output.

Key environmental fact: the `sgl_kernel` wheel installed here is a **gfx942 build** — `common_ops.so` embeds only `gfx942` code objects (24× `gfx942`, 0× `gfx1100`), and `sgl-kernel/setup_rocm.py:75` does `sys.exit(1)` for any target outside `{gfx942, gfx950}`. So on gfx1100 the op is *present* but runs wrong-ISA code.

## Finding 1 — `expert_ids` from `moe_align_block_size` are garbage (your question)

The HIP dispatcher is `sgl_kernel`-only (`fused_moe_triton/moe_align_block_size.py:14-15,76`; **no triton fallback**), and `expert_ids` is allocated with `torch.empty` (lines 66-68). Calling it with clean, in-range `topk_ids` (E=64, topk=6):

| | `num_tokens_post_padded` | `expert_ids` |
|---|---|---|
| correct (sort-by-expert) | 384–15488 | non-decreasing `0,0,…,1,1,…,63` |
| observed (M=512, block=128) | **0** | `[41,19,52,39,47,46,50,2,63,40,3,35,55,0,50,28,53,11,30,28,21,16,32,32,46,56,63,55,58,47,2,56,…]` |

`num_tokens_post_padded` returns **0** and the expert→block mapping is scrambled / non-monotonic: the op is non-functional. The OOB that *crashes* originates upstream — garbage `topk_ids` (as the equally-broken gfx942 `topk_softmax` emits) fed into `moe_align` yields `expert_ids` that are **all ≥ E**.

## Finding 2 — OOB `expert_ids` reproduces the exact crash

Injecting `expert_ids[0] = 2_000_000` into the real `fused_experts_impl` triton path (E=64, BF16):

```
Memory access fault by GPU node-2 (Agent handle: 0x4754a4e0) on address 0x7d3c4284c000.
Reason: Page not present or supervisor privilege.
Aborted (core dumped)            # SIGABRT, exit 134
```

Same fault class, same **GPU node-2**, same wording as the original report. Source confirms the mechanism: `fused_moe_kernel` loads `off_experts` (`fused_moe_triton_kernels.py:435`), its **only** guard is the `-1` sentinel (`:438`, gated on `filter_expert`), and the expert pointer is computed unmasked at `:469-473` (`b_ptr + off_experts * stride_be`). A garbage/OOB `off_experts` reads unmapped VRAM → page fault.

## Finding 3 — the triton kernel itself is correct on gfx1100

Same `fused_experts_impl` path, same triton kernel, only the `expert_ids` differ (correct in-range, monotonic): runs to completion, **no fault**, matches a torch MoE reference at **cosine 0.999992** (E=64, BF16). So the triton fused-MoE kernel compiles and computes correctly on RDNA3 given valid routing — the "triton GPU-hang" we saw earlier was a *symptom* of the garbage `expert_ids`, not a triton defect.

For completeness: feeding the real `sgl_kernel.moe_align_block_size` output through the same path is itself unstable — for some inputs it gives in-range-but-scrambled routing (→ **NaN** output, no crash); for others it yields an OOB `expert_id` that crashes the kernel directly. Either way the op is non-functional.

## Root cause (one level below `moe_align`)

On gfx1100 the **entire `sgl_kernel` MoE-gating/activation path is non-functional**, because the only buildable/available wheel is gfx942: `topk_softmax`, `moe_align_block_size`, `silu_and_mul`, `rms_norm`, `rotary_embedding` all miscompute or fault. The crash you identified is the OOB-`expert_ids` symptom of `moe_align` (fed garbage by `topk_softmax`); there is also a silent NaN-corruption mode.

## Workaround we shipped

We route BF16 MoE through SGLang's own torch-native `moe_forward_native` and force the gating/activation ops to their torch-native paths (`forward_hip → forward_native`). SGLang then serves `baidu/Unlimited-OCR` end-to-end on gfx1100 with coherent OCR (forward matches `model.infer`, cosine ≈ 1.0). Separately — and unrelated to this crash — there's a residual SGLang↔HF greedy bf16 divergence on ~12% of pages (paged-attention reduction order + bf16 accumulation); that's the known non-gfx1100 SGLang↔HF greedy-divergence issue, not an op bug.

## Reproduce

Self-contained script (`python verify_moe_gfx1100.py`) — drives the real triton path, controls only `expert_ids`:

<details><summary>verify_moe_gfx1100.py</summary>

```python
"""gfx1100 repro for sglang #30245. Drives the REAL fused_experts_impl triton
path; controls only expert_ids via a moe_align_block_size monkeypatch."""
import torch
import torch.nn.functional as F
import triton
import sglang.srt.layers.moe.fused_moe_triton.fused_moe as fm
import sglang.srt.server_args as _sa


class _D:                                   # fused_experts_impl's config path reads a few server-arg flags
    enable_deterministic_inference = False
    dp_size = 1
    tp_size = 1
    def __getattr__(self, k): return False


_sa._global_server_args = _D()
fm._orig = fm.moe_align_block_size          # the real wrapper (-> sgl_kernel) on this host


def alloc(t, bs, E):
    ntp = t.numel() * bs if t.numel() < E + 1 else t.numel() + (E + 1) * (bs - 1)
    sid = torch.empty((ntp,), dtype=torch.int32, device=t.device)
    eid = torch.empty((triton.cdiv(ntp, bs),), dtype=torch.int32, device=t.device)
    return sid, eid, torch.empty((1,), dtype=torch.int32, device=t.device)


def correct(t, bs, E):                       # ground-truth sort-by-expert + block padding
    sid, eid, ntp = alloc(t, bs, E)
    M, K = t.shape; NT = M * K; dev = t.device
    exp = t.reshape(-1).to(torch.int64); tok = torch.arange(NT, device=dev)
    o = torch.argsort(exp, stable=True); se, st = exp[o], tok[o]
    cnt = torch.bincount(se, minlength=E); wp = 0; eids = []; base = 0
    for e in range(E):
        c = int(cnt[e]); g = st[base:base + c]; base += c
        if not c: continue
        npd = (-c) % bs; sl = c + npd
        sid[wp:wp + c] = g.to(torch.int32)
        if npd: sid[wp + c:wp + sl] = NT
        eids += [e] * (sl // bs); wp += sl
    if wp < sid.numel(): sid[wp:] = NT
    ntp[0] = wp; ei = torch.tensor(eids, device=dev, dtype=torch.int32); eid.zero_(); eid[:ei.numel()] = ei
    return sid, eid, ntp


def make(mode):
    def f(t, bs, E):
        sid, eid, ntp = correct(t, bs, E)
        if mode == "oob": eid[0] = 2_000_000      # inject one OOB expert within num_tokens_post_padded
        return sid, eid, ntp
    return f


def moe_ref(hs, w1, w2, tw, ti):              # plain-PyTorch MoE reference
    M, H = hs.shape; E, N2, _ = w1.shape; I = N2 // 2
    out = torch.zeros(M, H, device=hs.device); hs, w1, w2 = hs.float(), w1.float(), w2.float()
    for m in range(M):
        for k in range(ti.shape[1]):
            e = int(ti[m, k]); g = F.linear(hs[m], w1[e])
            out[m] += float(tw[m, k]) * F.linear(F.silu(g[:I]) * g[I:], w2[e])
    return out


dev = "cuda"; E, H, inter, K, M = 64, 256, 512, 2, 64
g = torch.Generator(device=dev).manual_seed(0)
hs = (torch.rand(M, H, generator=g, device=dev) * 2 - 1).bfloat16()
w1 = (torch.rand(E, 2 * inter, H, generator=g, device=dev) * 2 - 1).bfloat16().contiguous()
w2 = (torch.rand(E, H, inter, generator=g, device=dev) * 2 - 1).bfloat16().contiguous()
ti = torch.randint(0, E, (M, K), generator=g, device=dev, dtype=torch.int32)
r = torch.rand(M, K, generator=g, device=dev); tw = (r / r.sum(1, keepdim=True)).bfloat16()
ref = moe_ref(hs, w1, w2, tw, ti)

for mode in ("valid", "oob"):
    fm.moe_align_block_size = make(mode)
    out = fm.fused_experts_impl(hs, w1, w2, tw, ti, activation="silu", is_gated=True)
    torch.cuda.synchronize()
    if mode == "valid":
        a, b = out.reshape(-1).float(), ref.reshape(-1).float()
        print(f"{mode}: cosine(vs torch ref)={torch.dot(a, b) / (a.norm() * b.norm() + 1e-9):.6f}  finite={bool(torch.isfinite(out).all())}")
    # mode == "oob": the unmasked tl.load(b_ptr + off_experts*stride_be) below faults -> SIGABRT
```

</details>

Expected output (verified on this host):

```
valid: cosine(vs torch ref)=0.999992  finite=True
Memory access fault by GPU node-2 ... Reason: Page not present or supervisor privilege.   # oob -> SIGABRT (exit 134)
```

## On the issue's asks

Ask #2 (gfx1100 in `setup_rocm.py`) is the right direction, but it needs the **full gating stack built + verified for RDNA3**, not just `moe_align`. The good news from Finding 3: since triton itself is correct on gfx1100, building `topk_softmax` + `moe_align_block_size` correctly for gfx1100 would unblock the fast triton fused-MoE path (our torch-native workaround is correct but considerably slower — a per-expert `F.linear` loop rather than a fused kernel). Happy to run a gfx1100 `sgl_kernel` build against this repro if it'd help.

**Comment by nicdumz** [comment]
@dondetir @WhatGhost would you have permissions to re-open? Lack of activity <> not useful or needed :-)
